### PR TITLE
Support unicode character for `initcap` function

### LINKED ISSUE #13711: Support unicode character for `initcap` function
### Is your feature request related to a problem or challenge?

Make `initcap` function work with Unicode character.

### Describe the solution you'd like

Current behavior of DataFusion:
```sql
select initcap('đẹp đẽ êm ả édouard'); --đẹP đẽ êM ả éDouard
select initcap('ñandú árbol éxito'); --ñAndú áRbol éXito
select initcap('олег иванович'); --олег иванович
select initcap('íslensku þjóðarinnar'); --íSlensku þJóðArinnar
select initcap('ελληνική επιστήμη'); --ελληνική επιστήμη
```
Expected behavior: (Postgres)
```sql
select initcap('đẹp đẽ êm ả édouard'); --Đẹp Đẽ Êm Ả Édouard
select initcap('ñandú árbol éxito'); --Ñandú Árbol Éxito
select initcap('олег иванович'); --Олег Иванович
select initcap('íslensku þjóðarinnar'); --Íslensku Þjóðarinnar
select initcap('ελληνική επιστήμη'); --Ελληνική Επιστήμη
```

### Describe alternatives you've considered

_No response_

### Additional context

Found in [comment](https://github.com/apache/datafusion/pull/13691#discussion_r1875885799) of @Dandandan

### PR DESCRIPTION
**Pull request by tlm365**
## Which issue does this PR close?

Closes #13711.

## What changes are included in this PR?
- Support UTF-8 for `initcap` function.
- Moved `initcap` from `string` module to `unicode` module.
- Update docs.

## Are these changes tested?
Yes, unit tests are included.

## Are there any user-facing changes?
Yes. `api change`.

### CODE REVIEW

File: datafusion/functions/src/unicode/initcap.rs [hunk 1]
```diff
@@ -74,7 +76,7 @@ impl ScalarUDFImpl for InitcapFunc {
             DataType::LargeUtf8 => make_scalar_function(initcap::<i64>, vec![])(args),
             DataType::Utf8View => make_scalar_function(initcap_utf8view, vec![])(args),
             other => {
-                exec_err!("Unsupported data type {other:?} for function initcap")
+                exec_err!("Unsupported data type {other:?} for function `initcap`")
```
Comment (2010YOUY01) [inline datafusion/functions/src/unicode/initcap.rs:79]:
```suggestion
                internal_err!("Unsupported data type {other:?} for function `initcap`")
```
Incompatible arg should be checked during planning before, thus here is unreachable, and we can use `internal_err` to indicate a potential bug if it's executed

File: datafusion/sqllogictest/test_files/string/string_query.slt.part [hunk 2]
```diff
@@ -460,7 +460,7 @@ Andrew Datafusion📊🔥
 Xiangpeng Datafusion数据融合
 Raphael Datafusionдатафусион
 Under_Score Un Iść Core
-Percent Pan Tadeusz Ma Iść W KąT
+Percent Pan Tadeusz Ma Iść W Kąt
```
Comment (2010YOUY01) [inline datafusion/sqllogictest/test_files/string/string_query.slt.part:463]: I've checked and this result is compatible with PostgreSQL

**Review by 2010YOUY01 — APPROVED** [review]
Thank you, this looks good to me

File: datafusion/functions/src/unicode/initcap.rs (hunk 1 above)
Comment (tlm365) [reply datafusion/functions/src/unicode/initcap.rs:79]: @2010YOUY01 Thanks for reviewing,

> Incompatible arg should be checked during planning before, thus here is unreachable, and we can use internal_err to indicate a potential bug if it's executed

Make sense to me👍

File: datafusion/functions/src/unicode/initcap.rs [hunk 3]
```diff
@@ -213,7 +243,7 @@ mod tests {
             Ok(Some("Hi Thomas")),
             &str,
             Utf8,
-            StringArray
+            StringViewArray
```
Comment (goldmedal) [inline datafusion/functions/src/unicode/initcap.rs:246]: 👍

File: datafusion/functions/src/unicode/initcap.rs [hunk 4]
```diff
@@ -123,24 +124,36 @@ fn get_initcap_doc() -> &'static Documentation {
...
+            builder.append_value(initcap_str);
+        }
+        None => builder.append_null(),
+    });
 
-    Ok(Arc::new(result) as ArrayRef)
+    Ok(Arc::new(builder.finish()) as ArrayRef)
 }
 
 fn initcap_string(input: Option<&str>) -> Option<String> {
```
Comment (goldmedal) [inline datafusion/functions/src/unicode/initcap.rs:159]:
```suggestion
fn initcap_string(input: &str) -> String {
```
It's only used by L135 and L150 and never be `None`. I think we can remove the `Option`.

File: datafusion/functions/src/unicode/initcap.rs [hunk 5]
```diff
@@ -149,13 +162,16 @@ fn initcap_string(input: Option<&str>) -> Option<String> {
...
+                }
             } else {
-                c.to_ascii_uppercase()
-            };
-            result.push(transformed);
-            prev_is_alphanumeric = c.is_ascii_alphanumeric();
+                for uc in c.to_uppercase() {
+                    result.push(uc);
+                }
+            }
```
Comment (goldmedal) [inline datafusion/functions/src/unicode/initcap.rs:173]: Does this change slow down the ASCII-only case? If yes, maybe we can consider keeping a faster path for ASCII-only cases. (like https://github.com/apache/datafusion/issues/12306)

**Review by goldmedal — APPROVED** [review]
Thanks @tlm365 It looks good to me but some minor comments.

File: datafusion/functions/src/unicode/initcap.rs (hunk 5 above)
Comment (tlm365) [reply datafusion/functions/src/unicode/initcap.rs:173]: @goldmedal thanks for reviewing

> Does this change slow down the ASCII-only case? If yes, maybe we can consider keeping a faster path for ASCII-only cases. (like https://github.com/apache/datafusion/issues/12306)

Seems like keeping a fast path for ASCII-only cases is beneficial. I will check and update it soon.

**Review by goldmedal — APPROVED** [review]

**Comment by goldmedal** [comment]
Thanks @tlm365 and @2010YOUY01 for reviewing

**Pull request merged**